code/android-core-library/src/main/java/com/adobe/marketing/mobile/EventCoder.java

```diff
@@ -28,184 +28,140 @@
+        if (eventString == null) {
+            return null;
+        }
+
+        try {
+            final JSONObject json = new JSONObject(eventString);
+            final String name = optString(json, NAME, null);
+            final String uniqueIdentifier = optString(json, UUID, null);
+            final String source = optString(json, SOURCE, null);
+            final String type = optString(json, TYPE, null);
+            final Map<String, Object> data = toMap(json.optJSONObject(DATA));
+            final long timestamp = json.optLong(TIMESTAMP, 0);
+            final String responseId = optString(json, RESPONSE_ID, null);
+            final JSONArray maskJsonArray = json.optJSONArray(MASK);
+
+            String[] mask = null;
+            if (maskJsonArray != null) {
+                mask = toList(maskJsonArray).toArray(new String[0]);
+            }
+
+            final Event ret = new Event.Builder(name, type, source, mask)
+                    .setUniqueIdentifier(uniqueIdentifier)
+                    .setTimestamp(timestamp)
+                    .setEventData(data)
+                    .setResponseId(responseId)
+                    .build();
+            return ret;
+        } catch (JSONException e) {
+            return null;
+        }
+    }
+
+    /**
+     * Encode an event to a json string
+     *
+     * @param event the event to encode
+     * @return json string represents all the fields of the event, otherwise returns null if the event is null or there is json error
+     */
+    public static String encode(final Event event) {
+        if (event == null) {
+            return null;
+        }
+
+        JSONObject json = new JSONObject();
+        try {
+            json.put(NAME, event.getName());
+            json.put(TYPE, event.getType());
+            json.put(SOURCE, event.getSource());
+            json.put(UUID, event.getUniqueIdentifier());
+            json.put(TIMESTAMP, event.getTimestamp());
+            json.put(DATA, JSONObject.wrap(event.getEventData()));
+            json.put(RESPONSE_ID, event.getResponseID());
+            json.put(MASK, JSONObject.wrap(event.getMask()));
+        } catch (JSONException e) {
+            return null;
+        }
+
+        return json.toString();
+    }
+
+    /**
+     * Returns the value mapped by {@code key} if it exists, or the {@code fallback}
+     * if no such mapping exists.
+     * Exists because {@code JSONObject#optString} does not allow a null fallback.
+     *
+     * @param jsonObject the json from which the key is to be fetched
+     * @param key the key that is to be fetched
+     * @param fallback the fallback value if the key does not exist in {@code jsonObject}
+     *
+     * @return value mapped by {@code key} if it exists in {@code jsonObject}, or {@code fallback}
+     * if no such mapping exists.
+     */
+    private static String optString(final JSONObject jsonObject, final String key, @Nullable final String fallback) {
```

Comment:
https://developer.android.com/reference/org/json/JSONObject#optString(java.lang.String,%20java.lang.String)

Reply:
The [sources from android 10 onwards](https://android.googlesource.com/platform/libcore/+/refs/heads/android10-release/json/src/main/java/org/json/JSONObject.java#580) have a NonNull  annotation for the fallback. I  see waring when using a null fallback because the target sdk for this project is 31.

Reply:
I missed reading the doc commet.